docs/proposals/subscriber-fan-in-fan-out.md

```diff
@@ -0,0 +1,189 @@
+
+Currently, when a subscriber is added, we take the existing channel or
+create one if not present and hand that channel to the subscriber.
+Instead, we would now have a registry of subscribers to topics:
+
+```go
+type WorkEngine struct {
+	subscribers map[WorkTopic][]WorkSubscriber
+	...
+} 
+
+```
+
+
+
+### Channel changes
+There a couple of options available as the goal here is to avoid
+any blocking:
+
+#### 1) Keep the single channel but remove potential for blocking
+
+Currently, the topic channel is handed to the single subscriber which
+performs all the work. As mentioned this means no more messages can be
+read from the channel until this unit of work is complete. A change to
+how the messages are handled could alleviate any potential issue. The
+reader would change to be the work engine which would start a read loop
+for each of the topics when the broker started and distribute incoming
+messages to any number of subscribers async:
+As long as the principal that subscribers shouldn't depend on each other
+is maintained there shouldn't be issues with this approach.
+```go 
+for{
+
+   select{
+       case msg := <-provisionChan:
+           for _,s := range e.subscribers[ProvisionTopic]{
+                go s(msg)              
+           }
+       case <-stop:
+       break
+   }
+}
+
+
+```
+
+The stop channel here would be tied to os signals
+[example](https://gist.github.com/reiki4040/be3705f307d3cd136e85).
+
+In this case, we would still have a single channel per topic, but as we
+hand off the message immediately and don't block waiting for the
+subscribers to finish their work, there is little to no chance for the
+channel to become full and so block the running jobs.
+
+#### Add work engine start method
+
+The work engine will now be the one reading messages from the channel
+and passing it on to the Observers. To do this I propose we add a  public
+``Start(stop <-chan struct{})`` method, that would be called in its own
+go routine during ASB start up. I also propose adding a signal channel
+that will send a message to anything running  in the background
+(could also use context.Context). This would signal for the background
+loops etc to stop.
+This start method would start a loop for all of the topics as shown
+above and close them down once a message was received.
+
+#### 2) Create a channel per job
+
+Remove the single channel per topic and replace it with a channel per
+job. The work engine would be updated to create a channel per Job
+when starting a new job. The work engine would be responsible for
+the lifecycle of these channels. As above, the subscribers would be
+registered in a map and called in a non-blocking way when a message was received.
+While more complex, this method would allow us to leverage more control
+(use a buffer to control how much could be sent while still
+not impacting on any other Jobs that running).
+
+Sudo Code:
+
+```go 
+
+StartNewJob(token string, work Work, topic WorkTopic){
```

Comment:
Just for my own clarification, the var `a` in the Sudo Code is for the work engine itself.

so from this, we see that a keeps track of all of the channels per a token, and keeps track of all of the subscribers for a given topic. Just want to make sure there is no other concerns that the engine is keeping track of.

Also, do we need to think about passing a context around here? I envision a use case where a subscriber says hey some really odd happened here and needs to send a message to other subscribers. I believe that is the use case for context?

Reply:
Yes those are the two things that would need to be tracked by the engine, If we were to use a channel per job.

On context, it is normally used to signal timeout and cancellations. I would not be against accepting a context as part of a Jobs definition to enable cancellation or potentially timeouts. 
Context can also carry values across API lines, but I am not such a fan of this approach.

Reply:
I think adding the context for timeout is something that we should do.

I think that canceling a running subscriber if another subscriber fails is something we should consider. 

@eriknelson @jmrodri do you guys have thoughts on timeouts or cancellations?

Reply:
> I think that canceling a running subscriber if another subscriber fails is something we should consider.

My gut feeling is that they're intended to take a simple data structure that contains the result of some work and do something useful with it completely independent from other subscribers. I'm not sure I like the idea of them communicating to each other.

Reply:
I think the concept of Jobs that can be cancelled is probably a proposal of its own and outside the remit of this work. 
If we had jobs that could somehow be cancelled, then the Job itself would take the context, I don't think the subscribers would as they only react to messages passed in. In the case of cancellation, the likelihood is a message would be passed in with a state of cancelled rather than the subscriber watching the context itself.